UI support cloud storage

Marishka17 opened the pull request:
<!---
Copyright (C) 2020-2021 Intel Corporation

SPDX-License-Identifier: MIT
-->

<!-- Raised an issue to propose your change (https://github.com/opencv/cvat/issues).
It will help avoiding duplication of efforts from multiple independent contributors.
Discuss your ideas with maintainers to be sure that changes will be approved and merged.
Read the [CONTRIBUTION](https://github.com/opencv/cvat/blob/develop/CONTRIBUTING.md)
guide. -->

<!-- Provide a general summary of your changes in the Title above -->

### Motivation and context
<!-- Why is this change required? What problem does it solve? If it fixes an open
issue, please link to the issue here. Describe your changes in detail, add
screenshots. -->
Dependent on [PR](https://github.com/openvinotoolkit/cvat/pull/3313)
Dependent on [PR](https://github.com/openvinotoolkit/cvat/pull/3386)
Related issue: #863
### How has this been tested?
<!-- Please describe in detail how you tested your changes.
Include details of your testing environment, and the tests you ran to
see how your change affects other areas of the code, etc. -->

### TODOs
- [x] fix search
- [x] cards with equal size
- [x] update query (when creating storage and go to the page of all storages there is no update, need update page manually)
- [x] add support for an access key, secret key pair for AWS S3 (instead of/or access key, secret key, session token set) 
- [x] clear task form after the task was created
- [x] cloud storage creating a page
- [x] cloud storage updating page
- [x] cloud storage listing page
- [x] new tab on creating task page
- [x] manually tested creating && updating cloud storages (S3 and credentials, S3 and anonymous access, Azure Blob container and credentials) && creating tasks with images from a cloud storage
- [x] correctly updating credentials
- [x] fix e2e test
- [x] modernize the structure of the data tree
- [x] fix all comments after team demonstration


### Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply.
If an item isn't applicable by a reason then ~~explicitly strikethrough~~ the whole
line. If you don't do that github will show incorrect process for the pull request.
If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [x] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

nmanovic commented:
![image](https://user-images.githubusercontent.com/40690625/129384687-0bfef121-c62c-4833-9e84-46c2aaae951d.png)

- [ ] Cansel -> Cancel
- [ ] Icon for AWS should be adjusted by the middle of the provider field
- [ ] When I click credentials and don't update them, probably it is intuitive to show "dummy" values (as in the beginning). Now they become empty always.
- [ ] Need to add some hint (or link to documentation) for "Manifests" field. Nobody knowns that it is.

nmanovic commented:
I tried to add the same storage with a different display name and got the error below. Why do you think we should prohibit that?

![image](https://user-images.githubusercontent.com/40690625/129401754-14e4bd77-9e78-4c94-b5da-ed30e80e0f37.png)

nmanovic commented:
- [ ] When a cloud storage is created, we need to check that the provided manifest files exist.

nmanovic commented:
- [ ] It will be great to distiguish between two typical cases: cannot find a file on the cloud storage and cannot access the storage with provided credentials. I deactivated my keys for the first cloud storage and got the following error when I try to access the storage:

![image](https://user-images.githubusercontent.com/40690625/129403398-582d9e95-ec73-4608-8200-780721dfa778.png)

nmanovic commented:
@Marishka17 , but overall the functionality works fine. Great job!

nmanovic commented:
@Marishka17 , when are you going to support `export task` feature?

![image](https://user-images.githubusercontent.com/40690625/132313727-37ed7d07-6b23-4b4a-9492-b7c5c53cfcb9.png)

nmanovic commented:
@Marishka17 ,

Select all files feature is missing and it can be pain if there are a lot of images in the root directory. 

![image](https://user-images.githubusercontent.com/40690625/132314591-7044f766-e376-4c5d-9bad-91fc60e1aa50.png)

nmanovic commented:
@Marishka17 , in many cases it is a good idea to choose first manifest file by default. I believe in many cases users will have only one. Now users have to choose a manifest explicitly.

![image](https://user-images.githubusercontent.com/40690625/132316489-6192c75e-3a2b-40d0-b562-1f8caf26393d.png)

nmanovic commented:
@Marishka17 , is it possible to adjust the icon below?

![image](https://user-images.githubusercontent.com/40690625/132316654-79bfadce-9daf-4d1a-b4bf-b3bc31dcc937.png)

nmanovic commented:
All mentioned problems are not critical. I will create a separate issue and describe them. Thus I'm fine with the PR.

@ActiveChooN , could you please review it ASAP?

Marishka17 commented:
> @Marishka17 , when are you going to support `export task` feature?

I think this should be implemented in a separate PR

ActiveChooN reviewed (commented):
Have few thoughts:
- It can be helpful to add `Select all` checkbox to the cloud storage files selection tab
And some questions not related to UI PR:
- Can manifest files be detected automatically? I guess just searching in the root directory and its subdirectories can cover most of the cases.
- Can a custom S3 server be selected, for example as it did in Cyberduck for example?
If I found something else I will let you.

Marishka17 commented:
> * Can manifest files be detected automatically? I guess just searching in the root directory and its subdirectories can cover most of the cases.

Surely this can be accomplished by filtering using the corresponding SDK, but manually specifying the required manifests was discussed in the last demo of the solution.

> * Can a custom S3 server be selected, for example as it did in Cyberduck for example?

I'm not really sure about this, I'll check

Marishka17 commented:
@nmanovic , @ActiveChooN 
- [x] default selected manifest on the task creating page
- [x] select all files on a task creating page

Marishka17 commented on cvat-ui/src/consts.ts:
```diff
@@ -25,6 +25,25 @@ const DEFAULT_PROJECT_SUBSETS = ['Train', 'Test', 'Validation'];
 const INTEL_TERMS_OF_USE_URL = 'https://www.intel.com/content/www/us/en/legal/terms-of-use.html';
 const INTEL_COOKIES_URL = 'https://www.intel.com/content/www/us/en/privacy/intel-cookie-notice.html';
 const INTEL_PRIVACY_URL = 'https://www.intel.com/content/www/us/en/privacy/intel-privacy-notice.html';
+const DEFAULT_AWS_S3_REGIONS: string[][] = [
```
I did not find an appropriate opportunity specifically for service s3, only for ec2, but I'm not sure if all services are supported in every region. Also, the current implementation has some flexibility, on the one hand, there is a list of available regions, on the other hand, the customer can expand it. For the first version, it was decided to keep the current approach.

Marishka17 commented on cvat-ui/src/components/file-manager/cloud-storages-files.tsx:
```diff
@@ -0,0 +1,211 @@
+            return nodes;
+        };
+        if (initialData.children && content) {
+            // TODO: need to optimizate this
```
I forgot to remove this

Marishka17 commented on cvat-ui/src/components/create-cloud-storage-page/cloud-storage-form.tsx:
```diff
@@ -195,7 +195,6 @@ export default function CreateCloudStorageForm(props: Props): JSX.Element {
 
         if (cloudStorage) {
             cloudStorageData.id = cloudStorage.id;
-            // TODO: it may be coincide with the real account name
```
This was my delusion. I checked the documentation and the fake data cannot coincide with the real ones because in real ones they can only be used lowercase letters

nmanovic commented:
@Marishka17 , probably an item to investigate after the PR is merged. A list of storages (I have only one) is loading longer than list of tasks (~10 tasks).

Marishka17 commented:
> @Marishka17 , probably an item to investigate after the PR is merged. A list of storages (I have only one) is loading longer than list of tasks (~10 tasks).

I think it might be related to getting storage status. I added making "dynamic" loading to do for the future, but for this, I would like to consult with @bsekachev  (by the way, I implemented this at the beginning, but it was not entirely correct), so I corrected this moment and did it by analogy with loading previews.

Marishka17 commented:
@aschernov  Hi, could I ask you to add documentation for this feature?

aschernov commented:
@Marishka17 , sure. We'll update the documentation

TOsmanov commented:
I found that the cloud storage element on the cloud storage page has an image that has the wrong cropping:
![image](https://user-images.githubusercontent.com/54434686/134636105-48e38a77-5662-4093-a1b8-22c6f909b4d7.png) 
if the page width <1366 px:
![image](https://user-images.githubusercontent.com/54434686/134636131-e737a366-21c3-4837-8198-8d84d0e73000.png)
The image is displayed after creating a task using this cloud storage. Similar items on the projects page are displayed correctly

I used "azur blob container" I managed to create a task using images, but I can't create a task using video. In process creating a task appears: 
"Could not fetch cloud storage data"

In the console:
![image](https://user-images.githubusercontent.com/54434686/134636166-bd49f509-fb4f-4675-9d48-824b820e6104.png)

Marishka17 commented:
> I found that the cloud storage element on the cloud storage page has an image that has the wrong cropping:

Thanks. Could you please create a issue?

> I used "azur blob container" I managed to create a task using images, but I can't create a task using video. 

For this functionality can be used only dataset with images
